I don't know if this is a common practice is cpython codebase, but my compiler complains of logical ops not being explicit enough. Personally I think it's silly.

```
/Users/user.surnam/cpython/Modules/_elementtree.c:3639:35: warning: '&&' within '||' [-Wlogical-op-parentheses]
        if (target->events_append && target->pi_event_obj || target->insert_pis) {
            ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~ ~~
/Users/user.surnam/cpython/Modules/_elementtree.c:3639:35: note: place parentheses around the '&&' expression to silence this warning
        if (target->events_append && target->pi_event_obj || target->insert_pis) {
                                  ^
```

--- a/Lib/test/test_xml_etree.py
+++ b/Lib/test/test_xml_etree.py
@@ -2984,35 +2984,35 @@ class TreeBuilderTest(unittest.TestCase):
         class TreeBuilderSubclass(ET.TreeBuilder):
             pass
 
-        xml = "<a>text<?pi1?><!-- comment --><?pi2?>tail</a>"
+        xml = "<a>text<?pi1?> <!-- comment -->\n<?pi2?>tail</a>"
         parser = ET.XMLParser(target=ET.TreeBuilder(insert_comments=True))
         parser.feed(xml)
         a = parser.close()
         self.assertEqual(a[0].text, ' comment ')
-        self.assertEqual(a[0].tail, 'tail')
-        self.assertEqual(a.text, "text")
+        self.assertEqual(a[0].tail, '\ntail')
+        self.assertEqual(a.text, "text ")
 
         parser = ET.XMLParser(target=TreeBuilderSubclass(insert_comments=True))
         parser.feed(xml)
         a = parser.close()
         self.assertEqual(a[0].text, ' comment ')
-        self.assertEqual(a[0].tail, 'tail')
-        self.assertEqual(a.text, "text")
+        self.assertEqual(a[0].tail, '\ntail')
+        self.assertEqual(a.text, "text ")
 
-        xml = "<a>text<!-- comment --><?pi data?>tail</a>"
+        xml = "<a>text<!-- comment -->\n<?pi data?>tail</a>"
         parser = ET.XMLParser(target=ET.TreeBuilder(insert_pis=True))
         parser.feed(xml)
         a = parser.close()
         self.assertEqual(a[0].text, 'pi data')
         self.assertEqual(a[0].tail, 'tail')
-        self.assertEqual(a.text, "text")
+        self.assertEqual(a.text, "text\n")
 
         parser = ET.XMLParser(target=TreeBuilderSubclass(insert_pis=True))
         parser.feed(xml)
         a = parser.close()
         self.assertEqual(a[0].text, 'pi data')
         self.assertEqual(a[0].tail, 'tail')
-        self.assertEqual(a.text, "text")
+        self.assertEqual(a.text, "text\n")
 
     def test_treebuilder_elementfactory_none(self):
         parser = ET.XMLParser(target=ET.TreeBuilder(element_factory=None))

--- a/Modules/_elementtree.c
+++ b/Modules/_elementtree.c
@@ -3636,7 +3636,7 @@ expat_pi_handler(XMLParserObject* self, const XML_Char* target_in,
         /* shortcut */
         TreeBuilderObject *target = (TreeBuilderObject*) self->target;
 
-        if (target->events_append && target->pi_event_obj || target->insert_pis) {
+        if ((target->events_append && target->pi_event_obj) || target->insert_pis) {
             pi_target = PyUnicode_DecodeUTF8(target_in, strlen(target_in), "strict");
             if (!pi_target)
                 goto error;

--- a/Modules/_elementtree.c
+++ b/Modules/_elementtree.c
@@ -2884,7 +2884,7 @@ treebuilder_handle_end(TreeBuilderObject* self, PyObject* tag)
 LOCAL(PyObject*)
 treebuilder_handle_comment(TreeBuilderObject* self, PyObject* text)
 {
-    PyObject* comment = NULL;
+    PyObject* comment;
     PyObject* this;
 
     if (treebuilder_flush_data(self) < 0) {
@@ -2923,7 +2923,7 @@ treebuilder_handle_comment(TreeBuilderObject* self, PyObject* text)
 LOCAL(PyObject*)
 treebuilder_handle_pi(TreeBuilderObject* self, PyObject* target, PyObject* text)
 {
-    PyObject* pi = NULL;
+    PyObject* pi;
     PyObject* this;
     PyObject* stack[2] = {target, text};
 
@@ -3532,8 +3532,8 @@ expat_end_ns_handler(XMLParserObject* self, const XML_Char* prefix_in)
 static void
 expat_comment_handler(XMLParserObject* self, const XML_Char* comment_in)
 {
-    PyObject* comment = NULL;
-    PyObject* res = NULL;
+    PyObject* comment;
+    PyObject* res;
 
     if (PyErr_Occurred())
         return;
@@ -3547,16 +3547,17 @@ expat_comment_handler(XMLParserObject* self, const XML_Char* comment_in)
             return; /* parser will look for errors */
 
         res = treebuilder_handle_comment(target,  comment);
+        Py_XDECREF(res);
+        Py_DECREF(comment);
     } else if (self->handle_comment) {
         comment = PyUnicode_DecodeUTF8(comment_in, strlen(comment_in), "strict");
         if (!comment)
             return;
 
         res = _PyObject_CallOneArg(self->handle_comment, comment);
+        Py_XDECREF(res);
+        Py_DECREF(comment);
     }
-
-    Py_XDECREF(res);
-    Py_DECREF(comment);
 }
 
 static void
@@ -3624,7 +3625,7 @@ static void
 expat_pi_handler(XMLParserObject* self, const XML_Char* target_in,
                  const XML_Char* data_in)
 {
-    PyObject* pi_target = NULL;
+    PyObject* pi_target;
     PyObject* data;
     PyObject* res;
     PyObject* stack[2];

